File: group 4/variant05/MembershipService.java
public final class MembershipService {
    public record Request(int months, int visits, boolean offPeak, boolean corporateClaimed) {}

    public record Result(String status, int months, int visits, int discountPercent, long entryFee, long total) {}

    public interface ClubCapacity {
        boolean acceptsMembers();
    }

    public interface CorporateRegistry {
        boolean isEligible();
    }

    public interface MembershipRepository {
        void save(Request request, Result result);
    }

    private final ClubCapacity capacity;
    private final CorporateRegistry corporate;
    private final MembershipRepository memberships;

    public MembershipService(ClubCapacity capacity, CorporateRegistry corporate, MembershipRepository memberships) {
        this.capacity = java.util.Objects.requireNonNull(capacity, "Залежність не задано: capacity");
        this.corporate = java.util.Objects.requireNonNull(corporate, "Залежність не задано: corporate");
        this.memberships = java.util.Objects.requireNonNull(memberships, "Залежність не задано: memberships");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.months(), 1, 12, "Строк абонемента");
        checkRange(request.visits(), 1, 29, "Кількість відвідувань");
        if (!capacity.acceptsMembers()) {
            return rejected("Немає доступних абонементів");
        }
        boolean eligible = request.corporateClaimed() && corporate.isEligible();
        int discount;
        if (request.offPeak() && eligible) {
            discount = 20;
        } else if (eligible) {
            discount = 15;
        } else if (request.offPeak()) {
            discount = 10;
        } else {
            discount = 0;
        }
        long monthly = 200_000 + Math.max(0, request.visits() - 12) * 5_000L;
        long base = request.months() * monthly;
        long entryFee = request.months() >= 11 ? 0 : 10_000;
        long total = base - base * discount / 100 + entryFee;
        Result result = new Result("Активовано", request.months(), request.visits(), discount, entryFee, total);
        memberships.save(request, result);
        return result;
    }

    private static Result rejected(String status) {
        return new Result(status, 0, 0, 0, 0, 0);
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}

File: group 1/variant05/membership_service.py
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Request:
    months: int
    visits: int
    off_peak: bool
    corporate_claimed: bool


@dataclass(frozen=True)
class Result:
    status: str
    months: int = 0
    visits: int = 0
    discount_percent: int = 0
    entry_fee: int = 0
    total: int = 0


class ClubCapacity(Protocol):
    def accepts_members(self) -> bool:
        ...

class CorporateRegistry(Protocol):
    def is_eligible(self) -> bool:
        ...

class MembershipRepository(Protocol):
    def save(self, request: Request, result: Result) -> None:
        """Фіксує всю операцію або повідомляє про помилку без часткових змін."""
        ...


def _range(value: int, minimum: int, maximum: int, field: str) -> None:
    if type(value) is not int or not minimum <= value <= maximum:
        raise ValueError("Недопустиме значення: " + field)


def _count(value: int, field: str) -> int:
    if type(value) is not int or value < 0:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


def _flag(value: bool, field: str) -> bool:
    if type(value) is not bool:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


class MembershipService:
    def __init__(self, capacity: ClubCapacity, corporate: CorporateRegistry, memberships: MembershipRepository) -> None:
        if capacity is None or corporate is None or memberships is None:
            raise ValueError("Усі залежності мають бути задані")
        self._capacity = capacity
        self._corporate = corporate
        self._memberships = memberships

    def process(self, request: Request) -> Result:
        if not isinstance(request, Request):
            raise ValueError("Заявку не задано або її формат некоректний")
        _range(request.months, 1, 12, "Строк абонемента")
        _range(request.visits, 1, 29, "Кількість відвідувань")
        if type(request.off_peak) is not bool:
            raise ValueError("Недопустима логічна ознака: off_peak")
        if type(request.corporate_claimed) is not bool:
            raise ValueError("Недопустима логічна ознака: corporate_claimed")
        if not _flag(self._capacity.accepts_members(), "Прийом клієнтів"):
            return Result("Немає доступних абонементів")
        corporate = request.corporate_claimed and _flag(self._corporate.is_eligible(), "Корпоративна пільга")
        if request.off_peak and corporate:
            discount = 20
        elif corporate:
            discount = 15
        elif request.off_peak:
            discount = 10
        else:
            discount = 0
        monthly = 200_000 + max(0, request.visits - 12) * 5_000
        base = request.months * monthly
        entry_fee = 0 if request.months >= 11 else 10_000
        total = base - base * discount // 100 + entry_fee
        result = Result("Активовано", request.months, request.visits, discount, entry_fee, total)
        self._memberships.save(request, result)
        return result
